<?xml version="1.0"?>
<!--
MIT License

Copyright (c) 2022 Gotelli Andrea

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-->



<launch>

    <!--    Position and angle of the limb  -->
    <arg name="x"   default="0.0"/>
    <arg name="y"   default="0.0"/>
    <arg name="yaw" default="0.0"/>

    <!-- Orientation of the attach point -->
    <arg name="a_roll"  default="0.0"/>
    <arg name="a_pitch" default="0.0"/>
    <arg name="a_yaw"   default="0.0"/>

    <!--    Number of the limb for creating an unique ID    -->
    <arg name="limb_number" default="0"/>

    <arg name="load_control" default="true"/>

    <!--    Type of base joint  -->
    <arg name="base_joint_type" default="revolute"/>

    <group ns="limb$(arg limb_number)">
        <!-- Load the URDF into the ROS Parameter Server -->
        <param name="robot_description"  command="$(find xacro)/xacro '$(find cpr_description)/urdf/limb_base_Till.xacro' x:=$(arg x)
                                                                                                                                              y:=$(arg y)
                                                                                                                                              yaw:=$(arg yaw)
                                                                                                                                              a_roll:=$(arg a_roll)
                                                                                                                                              a_pitch:=$(arg a_pitch)
                                                                                                                                              a_yaw:=$(arg a_yaw)" />

        <!-- Run a python script to the send a service call to gazebo_ros to spawn a URDF robot -->
        <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
                args="-urdf -model limb$(arg limb_number) -param robot_description"/>

        <!-- Load joint controller configurations from YAML file to parameter server -->
        <rosparam file="$(find cpr_description)/config/$(arg base_joint_type)_joint_config.yaml" command="load" if="$(eval arg('load_control'))"/>

        <param name="is_extendible" type="bool" value="true" unless="$(eval arg('load_control'))"/>

    </group>

</launch>
